### PR TITLE
[Permissions] make perms job cleanup worker interval configurable

### PR DESCRIPTION
**Pull request by thenamankumar**
Make `periodicGoroutine` accept the `getInterval` func to enable having a dynamic interval. Use that to integrate the perm syncer jobs cleaner with site config.

https://www.loom.com/share/a57d0441ea0040c89648a04775c250d3

## Test plan
- update site config
- check for changes in cleaner schedule.

### CODE REVIEW

File: internal/goroutine/periodic.go [hunk 1]
```diff
@@ -100,10 +103,14 @@ func NewPeriodicGoroutine(ctx context.Context, name, description string, interva
...
-	return newPeriodicGoroutine(ctx, name, description, interval, handler, operation, glock.NewRealClock())
+	return newPeriodicGoroutine(ctx, name, description, interval, handler, operation, glock.NewRealClock(), nil)
+}
+
+func NewPeriodicGoroutineWithMetricsAndDynamicInterval(ctx context.Context, name, description string, handler Handler, getInterval getIntervalFunc, operation *observation.Operation) *PeriodicGoroutine {
+	return newPeriodicGoroutine(ctx, name, description, getInterval(), handler, operation, glock.NewRealClock(), getInterval)
 }
 
-func newPeriodicGoroutine(ctx context.Context, name, description string, interval time.Duration, handler Handler, operation *observation.Operation, clock glock.Clock) *PeriodicGoroutine {
+func newPeriodicGoroutine(ctx context.Context, name, description string, interval time.Duration, handler Handler, operation *observation.Operation, clock glock.Clock, getInterval getIntervalFunc) *PeriodicGoroutine {
```
Comment (mrnugget) [inline internal/goroutine/periodic.go:113]:
```suggestion
func newPeriodicGoroutine(ctx context.Context, name, description string, getInterval getIntervalFunc, handler Handler, operation *observation.Operation, clock glock.Clock) *PeriodicGoroutine {
```

Why not do this and pass in `func() time.Time { return interval}` in `NewPeriodicGoroutineWithMetrics`?

**Review by mrnugget — APPROVED** [review]

File: schema/site.schema.json [hunk 2]
```diff
@@ -959,6 +959,11 @@
       "minimum": 5,
       "!go": { "pointer": true }
     },
+    "permissions.syncJobCleanupInterval": {
+      "description": "Time interval (in seconds) of how often cleanup worker should remove old jobs from permissions sync jobs table.",
+      "type": "integer",
+      "default": 60
```
Comment (sashaostrikov) [inline schema/site.schema.json:965]:
```suggestion
      "default": 60,
      "minimum": 1
```
To defend from cool hackers with their negative values

**Review by sashaostrikov — APPROVED** [review]
LGTM!

File: schema/site.schema.json (hunk 2 above)
Comment (thenamankumar) [reply schema/site.schema.json:965]: its taken care of in the code, but added here as well.

**Comment by thenamankumar** [comment]
DO NOT MERGE

File: enterprise/cmd/worker/main.go [hunk 3]
```diff
@@ -3,8 +3,10 @@ package main
 import (
 	"github.com/sourcegraph/sourcegraph/enterprise/cmd/sourcegraph/enterprisecmd"
 	"github.com/sourcegraph/sourcegraph/enterprise/cmd/worker/shared"
+	"github.com/sourcegraph/sourcegraph/internal/conf"
 )
 
 func main() {
 	enterprisecmd.DeprecatedSingleServiceMainEnterprise(shared.Service)
+	conf.Init()
```
Comment (efritz) [inline enterprise/cmd/worker/main.go:11]: `DeprecatedSingleServiceMainEnterprise` blocks so the `conf.Init()` won't be called.

Comment (thenamankumar) [reply enterprise/cmd/worker/main.go]: fixed, thanks for the help.

File: enterprise/cmd/frontend/worker/auth/perms_syncer_cleaner.go [hunk 4]
```diff
@@ -46,19 +89,26 @@ func (p *permissionSyncJobCleaner) Routines(_ context.Context, observationCtx *o
 		Metrics: m,
 	})
 
+	watchConfOnce.Do(func() {
+		loadCleanupIntervalFromConf()
```
Comment (efritz) [inline enterprise/cmd/frontend/worker/auth/perms_syncer_cleaner.go:93]:
```suggestion
```

conf.Watch will block until it calls its function argument the first time

File: internal/goroutine/periodic.go (hunk 1 above)
Comment (thenamankumar) [reply internal/goroutine/periodic.go]: updated

File: enterprise/cmd/frontend/worker/auth/perms_syncer_cleaner.go [hunk 5]
```diff
@@ -30,6 +30,16 @@ func (p *permissionSyncJobCleaner) Config() []env.Config {
 	return nil
 }
 
+var cleanupInterval = 60 * time.Second
+
+func loadCleanupIntervalFromConf() {
+	seconds := conf.Get().PermissionsSyncJobCleanupInterval
+	if seconds <= 0 {
+		cleanupInterval = 60 * time.Second
```
[the remark this answers is not in the document]
Comment (thenamankumar) [reply enterprise/cmd/frontend/worker/auth/perms_syncer_cleaner.go]: updated

**Review by efritz — APPROVED** [review]

File: enterprise/cmd/frontend/worker/auth/perms_syncer_cleaner.go [hunk 6]
```diff
@@ -30,6 +31,48 @@ func (p *permissionSyncJobCleaner) Config() []env.Config {
...
+	fmt.Printf("seconds: %d\n", seconds)
+	fmt.Printf("seconds: %d\n", seconds)
+	fmt.Printf("seconds: %d\n", seconds)
+	fmt.Printf("seconds: %d\n", seconds)
+	fmt.Printf("seconds: %d\n", seconds)
+	fmt.Printf("seconds: %d\n", seconds)
+	fmt.Printf("seconds: %d\n", seconds)
+	fmt.Printf("seconds: %d\n", seconds)
+	fmt.Printf("seconds: %d\n", seconds)
+	fmt.Printf("seconds: %d\n", seconds)
```
[the remark this answers is not in the document]
Comment (Strum355) [reply enterprise/cmd/frontend/worker/auth/perms_syncer_cleaner.go:68]: this is so me 💅

[the remark this answers is not in the document]
Comment (thenamankumar) [reply enterprise/cmd/frontend/worker/auth/perms_syncer_cleaner.go]: removed

File: enterprise/cmd/frontend/worker/auth/perms_syncer_cleaner.go (hunk 4 above)
Comment (thenamankumar) [reply enterprise/cmd/frontend/worker/auth/perms_syncer_cleaner.go]: removed

Comment (thenamankumar) [reply enterprise/cmd/frontend/worker/auth/perms_syncer_cleaner.go]: and tested again. works

**Pull request merged**